### PR TITLE
Added some comments and fixed one issue with the sea level

### PR DESCRIPTION
**Pull request by Qxji**
Also simplified that giant else if statement with chatgpt :D

### CODE REVIEW

**Comment by Qxji** [comment]
Also: I can't seem to get the drag and drop to work, but I didn't want to look into that atm.

**Comment by Qxji** [comment]
I realized when splitting of the dot the int() can't round properly, but I think it's better that way. It probably won't be noticeable anyways, but its probably better if it always rounds down, because if it rounds up there is a chance it actually overlaps with stuff it shouldn't..

File: city_data.py [hunk 1]
```diff
@@ -20,7 +20,9 @@ def load_xml(self, path2xml):
 
     # load terrain data from root
     def get_terrains(self):
-        sea_level = int(self.city_root.find('SeaLevel').text)
+        # apparently the SeaLevel is sometimes not a natural number, e.g. "207.328",
```
Comment (Qxji) [inline city_data.py:23]: See my comment in Conversation. The int() can't round properly but I think its fine.

**Comment by KevinUTAT** [comment]
Thank you for fixing my mistake and tidying up my messy code. :)
If you are actively using this project, I would love your feedback.
If you have a feature suggestion (or found bugs), please do not hesitate to raise issues (or PR, lol)

**Comment by KevinUTAT** [comment]
> I realized when splitting of the dot the int() can't round properly, but I think it's better that way. It probably won't be noticeable anyways, but its probably better if it always rounds down, because if it rounds up there is a chance it actually overlaps with stuff it shouldn't..

it only works by dragging a .cslmap file into the viewer area. I actually didn't implement this feature, cames from this [PR](https://github.com/KevinUTAT/CSL-QMapViewer/pull/2)

**Comment by Qxji** [comment]
> it only works by dragging a .cslmap file into the viewer area. I actually didn't implement this feature, cames from this [PR](https://github.com/KevinUTAT/CSL-QMapViewer/pull/2)

Yeah, the person who made that also made some perfomance improvements in their fork, might want to look into that. I am currently looking if I can't use your city_data code (you did the hard part) to generate an SVG of the city. I think vectors would work better than rasterisation, maybe later we can add that as an export option.

I am amazed at how much work you did with this project.

For some feedback: 
1) The original has much nicer graphics. It probably shouldn't be hard to change the style a bit so it looks better. Will look into that this weekend too.
2) Tunnels are currently not implemented, I'll have to look at the xml file and see if there is a tunnel thingy.
3) I have no experience with that, but is there a way to intruduce multiprocessing? Loading the stuff (segments, nodes, etc) in city_data all at the same time should already make the perfomance much better. Maybe the loading processes in each of the functions can also be parallelized?
4) I think its already on your To-Do but adding toggles for objects on the map - there has to be some way so that the map doesnt have to be redrawn everytime something is toggled - else there needs to be some big perfomance increases in that too.

So the biggest thing really is perfomance. And that is the most effort.

**Comment by Qxji** [comment]
I did some research, and apparently there are faster parsing methods for the xml file: One can use lxml with xpath expressions, and that apparently is faster then the normal methods. Would have to try that out tho.

**Comment by KevinUTAT** [comment]
> > it only works by dragging a .cslmap file into the viewer area. I actually didn't implement this feature, cames from this [PR](https://github.com/KevinUTAT/CSL-QMapViewer/pull/2)
> 
> Yeah, the person who made that also made some perfomance improvements in their fork, might want to look into that. I am currently looking if I can't use your city_data code (you did the hard part) to generate an SVG of the city. I think vectors would work better than rasterisation, maybe later we can add that as an export option.
> 
> I am amazed at how much work you did with this project.
> 
> For some feedback:
> 
> 1. The original has much nicer graphics. It probably shouldn't be hard to change the style a bit so it looks better. Will look into that this weekend too.
> 2. Tunnels are currently not implemented, I'll have to look at the xml file and see if there is a tunnel thingy.
> 3. I have no experience with that, but is there a way to intruduce multiprocessing? Loading the stuff (segments, nodes, etc) in city_data all at the same time should already make the perfomance much better. Maybe the loading processes in each of the functions can also be parallelized?
> 4. I think its already on your To-Do but adding toggles for objects on the map - there has to be some way so that the map doesnt have to be redrawn everytime something is toggled - else there needs to be some big perfomance increases in that too.
> 
> So the biggest thing really is perfomance. And that is the most effort.

Yeah, ability to export is probably be one of the priority next.
Response to the feedback:
1. You are absolutely right. I try to keep the same colour coding as the original viewer. But i can't seems to make the height to colour mapping to feel just right (the background). I encourage you to play around with it [here](https://github.com/KevinUTAT/CSL-QMapViewer/blob/main/terrains.py#L75)
2. Yes, every `Node` has property of `ug`(underground) and `og`(overground). All the node information is parsed already in `city_data.node_dict`, but not being used. So each `Segment` has start node and end node, you can use that to figure out if its tunnel/bridge, here is a rough example:
```cpp
start_node_is_tunnel = city_data.node_dict[current_segment.sn].ug
```
3. Yes, using threads would probably speed up the loading. I am personally not concerned about the loading speed, as user are usually patient with a one-time loading (as long as there is a loading bar, lol). The more important matric is the real time performance when panning and zooming. 
4. Yes, the empty panel on the right side is supposed to be a toggle list.

**Comment by KevinUTAT** [comment]
I am going to keep a to-dos in the issue panel with a label of `feature_request`, feel free to add

**Pull request merged**